### PR TITLE
Signal use

### PR DESCRIPTION
**Pull request by wild-endeavor**
# TL;DR
This PR adds some Flyte remote constructs around gate nodes so that pyflyte run works.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
These are the workflows that this was tested on.
https://github.com/flyteorg/flytesnacks/blob/gate-example/cookbook/core/control_flow/human_in_the_loop.py

### pyflyte
```
pyflyte run --remote core/control_flow/human_in_the_loop.py mainwf --a 3
pyflyte run --remote core/control_flow/human_in_the_loop.py wf_dyn --a 3
```

### Remote usage

Some sample usage of new remote bits
```
In [1]: from flytekit.remote.remote import FlyteRemote
   ...: from flytekit.configuration import Config
   ...: r = FlyteRemote(
   ...:     Config.auto(config_file="/Users/ytong/.flyte/dev.yaml"),
   ...:     default_project="flytesnacks",
   ...:     default_domain="development",
   ...: )

In [2]: r.list_signals("atc526g94gmlg4w65dth")
Out[2]:
[id {
  signal_id: "my-signal-name"
  execution_id {
    project: "flytesnacks"
    domain: "development"
    name: "atc526g94gmlg4w65dth"
  }
}
type {
  simple: BOOLEAN
}
, id {
  signal_id: "my-signal-name-2"
  execution_id {
    project: "flytesnacks"
    domain: "development"
    name: "atc526g94gmlg4w65dth"
  }
}
type {
  simple: INTEGER
}
]

In [3]: r.set_signal("my-signal-name", "atc526g94gmlg4w65dth", True)
In [4]: r.set_signal("my-signal-name-2", "atc526g94gmlg4w65dth", 5)
```

## Tracking Issue
https://github.com/flyteorg/flyte/issues/3125
https://github.com/flyteorg/flyte/issues/208

### CODE REVIEW

**Review by hamersaw — COMMENTED** [review]
one small comment nit

File: flytekit/clients/raw.py [hunk 1]
```diff
@@ -406,6 +409,20 @@ def get_task(self, get_object_request):
...
+    def set_signal(self, signal_set_request: SignalSetRequest) -> SignalSetResponse:
+        """
+        This sets a signal
+        """
+        return self._signal.SetSignal(signal_set_request, metadata=self._metadata)
+
+    @_handle_rpc_error(retry=True)
+    def list_signals(self, signal_list_request: SignalListRequest) -> SignalList:
+        """
+        This sets a signal
```
Comment (pingsutw) [inline flytekit/clients/raw.py:422]:
```suggestion
        This lists signals
```

File: tests/flytekit/unit/core/test_gate.py [hunk 2]
```diff
@@ -290,3 +291,36 @@ def cond_wf(a: int) -> float:
...
+
+    @task
+    def t2(a: int) -> int:
+        return a + 6
+
+    @workflow
+    def wf(a: int) -> typing.Tuple[int, int, int]:
+        zzz = sleep(timedelta(seconds=10))
+        x = t1(a=a)
+        s1 = wait_for_input("my-signal-name", timeout=timedelta(hours=1), expected_type=bool)
```
Comment (pingsutw) [inline tests/flytekit/unit/core/test_gate.py:309]: qq: should the name of the gate node be unique?

Comment (hamersaw) [reply tests/flytekit/unit/core/test_gate.py:309]: so this is actually an interesting question. signals ids are unique within a workflow execution. so if you had this same statement twice it would just wait twice for a single signal with the given name. nothing would break unless the expected types are different. presumably you could have two subworkflows that both wait on the same signal by using the same name. at least that's how it works in the backend. not sure if we want to explicitely restrict this from flytekit side?

Comment (pingsutw) [reply tests/flytekit/unit/core/test_gate.py:309]: got it, thanks for the explain

Comment (wild-endeavor) [reply tests/flytekit/unit/core/test_gate.py:309]: yes.

**Review by hamersaw — APPROVED** [review]

**Review by pingsutw — APPROVED** [review]

**Review by pingsutw — APPROVED** [review]
